### DISCUSSION TITLE
How to define "endpoint" without using "decorator syntax" ?

### DISCUSSION THREAD

**Discussion by 4n1qz5skwv** [litestar-org/litestar#1624] [Q&A] [answered]
An starlette example,

```python
async def mail_page(request: Request):
    ...

app = Starlette(routes=[
        Mount(root_path, routes=[
            Route(path='/{token:uuid}', endpoint=mail_page, methods=['GET','POST'], name='mail'),
        ])
    ]
)
```

This way we can have central control on the route connections and easy to configure.

**Comment by provinzkraut** [comment]
This is already how the decorators work. They don’t register a route handler, they just define it. They’re registered by passing it to the app, exactly like oh your example.

i recommend you read the documentation: https://docs.litestar.dev/latest/

**Reply by provinzkraut** [reply]
You’re not paying attention to the docs.

The example you showed is *not* what is happening. The decorators are *not* methods on the application instance. You still have to pass the route handlers to the application, which is exactly what you’re asking to do.

Yes, a decorator is used, but **the decorator does not register a route handler.**

**Reply by 4n1qz5skwv** [reply]
![Capture](https://user-images.githubusercontent.com/76768193/236666321-8c83b58d-2846-4156-9e64-dc61aed10e63.PNG)


Is it possible in Litestar to make "routes" without using "decorator syntex" ?

**Reply by provinzkraut** [reply, marked as answer]
The issue you're having is that you want to do this

> This way we can have central control on the route connections and easy to configure.

but arrived at the conclusion that the way to achieve it is to not use decorators.

This conclusion however is wrong, since what you want is already how Litestar works.

The equivalent of this Starlette app

```python
async def mail_page(request: Request):
    ...

app = Starlette(routes=[Route(path='/{token:uuid}', endpoint=mail_page, methods=['GET'], name='mail')])
```

is this Litestar app:

```python
@get("/{token:uuid}", name="mail")
async def mail_page(request: Request):
    ...

app = Litestar(route_handlers=[mail_page])
```

So to answer your question: You can't. Decorators are how you define the route handlers. 
But you also don't *need* to, since the decorators are not registering the route handlers at the same time. These are strictly separated in Litestar.

The sanic example you provided is *not* equivalent to what's happening in Litestar.

```python
@app.route("/test", methods=["GET")
async def handler(request):
  ...
```

in this Sanic example, `@app.route` does two things: It creates an endpoint function, and *registers it on the application*, which, going off your initial post here is what you want avoid. But this is not what's happening with Litestar's decorators. Notice how in the Sanic example `.route` is a **method on the application instance**. In Litestar, this is not the case. Decorating a function with a route handler decorator such as `@post` does not do anything besides creating a route handler. It does not register it, since there's no application associated with it.


You can even make the Litestar and Starlette equivalents look almost identical if you really want to:

```python
async def mail_page(request: Request):
    ...

mail_page = Route(path='/{token:uuid}', endpoint=mail_page, methods=['GET'], name='mail')

app = Starlette(routes=[mail_page])
```

is this Litestar app:

```python

async def mail_page(request: Request):
    ...

mail_page = get("/{token:uuid}", name="mail")(mail_page)

app = Litestar(route_handlers=[mail_page])
```

It's simply a difference in syntax, not in functionality.

**Reply by 4n1qz5skwv** [reply]
> mail_page = get("/{token:uuid}", name="mail")(mail_page)

THIS... I was asking if Litestar has some built-in option for that, instead of just using the decorator like this. I am already doing this. since this is a decorator. The point of this thread was to learn if I am missing any builtin functionality. Point of sanic example is to show how they have both "decorator" and "non-decorator" system. Since u have indeed use the decorator like this.. confirms that there is no "non-decorator" system litestar. So I marked that as a answer.

**Reply by provinzkraut** [reply]
> Point of sanic example is to show how they have both "decorator" and "non-decorator" system.

Yes, but they do so because their decorators do something else. They give you a "non-decorator" option so you can separate the definition of your endpoints from the registration. This however is *already the case* in Litestar. 

The only thing you'd gain from a non-decorator option in Litestar would be not using decorators. There's no functionality to be gained here. 

Let me ask you again: *Why* do you want to do this without decorators? What are you trying to achieve with that?

**Reply by 4n1qz5skwv** [reply]
> Let me ask you again: _Why_ do you want to do this without decorators? What are you trying to achieve with that?

![Capture](https://user-images.githubusercontent.com/76768193/236667111-7ccd21d5-9745-411e-bdeb-37ed23815d80.PNG)

It's a code from a tiny app. Just look at this, isn't this beautiful. You have total control, overview, easy to maintain code right there. 
If I want to rename a route, I don't have to track down the function and edit decorator, If i want to change the function, I don't have to Cut  and paste "decorator code" from one function to another. It's like a Tree. A tree, where every branches I can reach and edit without taking a step.

**Reply by Goldziher** [reply]
You can use the `route` decorator to get a similar effect to a degree.

You do realize though what you're asking for is purely aesthetic?

**Reply by 4n1qz5skwv** [reply]
It's just that, all the lib I used before (before i found Litestar) has "non-decorator" system or both. Since Starlite is changing to Litestar, I thought may be It's not on the new doc yet, So I made this thread to confirm.

> purely aesthetic?

Really though? there a reason almost all the lib has this option first and then the decorator maybe for the convenience.

**Reply by provinzkraut** [reply]
> It's a code from a tiny app. Just look at this, isn't this beautiful

Well, I'd say this is very much limited to small apps, for which the "tracking down code" part should be relatively easy already.

Managing hundreds or thousands of routes like this isn't feasible anymore, and you'd usually organise things into their own submodules, routers, etc. 

There's also a lot more going on in Litestar's handler decorators than Starlette's decorators, so if you start adding 10+ keyword arguments to each of those, it's going to be quite a mess. 

Your example also only makes sense if you're operating under the assumption that Litestar works the same way Starlette does, which it does not. If you were to add functionality like this to Litestar, it would make things much harder to reason about, and you'd end up doing a lot more "tracking down".

**A simple example**

Let's assume this is your endpoint function:

```python
async def handler(name: str, limit: int, user: User) -> str:
   ...

```

and then you add it to your application, which lives in a different file, like this

```python
from .routes import handler

app = Litestar(
  route_handlers=[
    Route(
      path="/{name:str}", 
      dependencies={"user": Provide(get_user)},
      guards=[authenticate_user],
      methods=["GET"],
      cache=True,
      raises=[NotAuthorized],
      media_type="text/xml"
    )
  ]
)
``` 

when you now look at the `handler` function, you know nothing about how it behaves, or how it *should* behave; What's the source of the parameters? What data type should it return? Is it cached and for how long? What exceptions should it raise? You'd have to track down all the places this function is potentially used in, and check there. This is not a great developer experience.

Moreover, if you wanted to change the `name` parameter, you wouldn't know if you could easily do that. You'd first have to check where it comes from, and then change the route definition in the central location.

So what's the solution to over come this?
Put all the relevant information close to where it's being used.

```python
async def handler(name: str, limit: int, user: User) -> str:
   ...

handler_route = Route(
    path="/{name:str}", 
    dependencies={"user": Provide(get_user)},
    guards=[authenticate_user],
    methods=["GET"],
    cache=True,
    raises=[NotAuthorized],
    media_type="text/xml"
)
```

and then use it in the app like so

```python
from .routes import handler_route

app = Litestar(route_handlers=[handler_route])
```

which is exactly how you would solve the problem of information locality in Starlette. But because this is not that easy on the eyes and Python already has a syntax that's specifically made for this very purpose, instead, we just use a decorator:

```python
@get(
  "/{name:str}",
  dependencies={"user": Provide(get_user)},
  guards=[authenticate_user],
  methods=["GET"],
  cache=True,
  raises=[NotAuthorized],
  media_type="text/xml"
)
async def handler(name: str, limit: int, user: User) -> str:
  ...
```

which is then again registered like this in the application


```python
from .routes import handler

app = Litestar(route_handlers=[handler])
```

**Reply by provinzkraut** [reply]
> Really though? there a reason almost all the lib has this option first and then the decorator maybe for the convenience.

Yes, and I explained the reason to you a few times now. It's because those decorators do a different thing than the ones in Litestar. There is a *functional* difference there. This functional difference is not present in Litestar.
